velox/type/Filter.cpp

```diff
@@ -458,6 +458,71 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kMultiRange: {
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+      // TODO: verify nanAllowed with the other filter.
+      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+      std::vector<std::unique_ptr<Filter>> merged;
+      for (auto const& filter : this->filters()) {
+        for (auto const& filterOther : multiRangeOther->filters()) {
+          auto innerMerged = filter->mergeWith(filterOther.get());
+          switch (innerMerged->kind()) {
+            case FilterKind::kAlwaysFalse:
+            case FilterKind::kIsNull:
+              break;
+            default:
+              merged.push_back(std::move(innerMerged));
+          }
+        }
+      }
+
+      if (merged.empty()) {
+        return nullOrFalse(bothNullAllowed);
+      } else if (merged.size() == 1) {
+        // If both sides allow null then make sure the only filter matches with
```

Comment:
Shouldn't we return merged[0] with nullAllowed set to bothNullAllowed?

([1.2, 3.4] OR [4.5, 6.7]) AND ([5.0, 6.0] OR [7.5, 10.0]) => [5.0, 6.0]

Reply:
Now I see your suggestion: we need a clone but allow override `nullAllowed`